pkgs/tools/archivers/unzip/implicit-declarations-fix.patch

```diff
@@ -0,0 +1,22 @@
+--- a/unix/configure	2009-04-16 15:25:12.000000000 -0400
++++ b/unix/configure	2023-05-30 15:18:33.670321348 -0400
+@@ -408,7 +408,7 @@
+ echo Check for errno declaration
+ cat > conftest.c << _EOF_
+ #include <errno.h>
+-main()
++int main()
+ {
+   errno = 0;
+   return 0;
+@@ -419,6 +419,9 @@
+ 
+ echo Check for directory libraries
+ cat > conftest.c << _EOF_
++typedef struct _dir DIR;
++DIR* opendir(const char*);
++int closedir(DIR*);
```

Comment:
I think the correct fix is to include the headers that define it:

```
#include <sys/types.h>
#include <dirent.h>
```

Maybe we could borrow the fix from other distributions?

Reply:
My reasoning for not using the headers here is that this particular test is determining which library has those symbols. It checks which headers to include later. I was trying to preserve that original separation of checks.

In practice, it shouldn’t matter for the platforms nixpkgs supports. If you’d rather I not bother, I can push an updated patch that just includes those headers.

Reply:
I don't have a strong opinion for this case.

Generally it's easy to cause unintended breakage by defining slightly incompatible prototypes (macros vs symbols, `struct` mismatch, `struct fs integral typedef`, `-static -flto` seeing through struct definitions). Would be unfortunate if it comes back and bites us later on more exotic target or toolchain.

Reply:
I agree it’s a bad test. It just checks for the first library that happens not to have missing symbols when you try to link it. I went ahead and pushed a commit to use the headers.